runtime/src/main/scala/grpc/akkastreams/GrpcGraphStage.scala

```diff
@@ -0,0 +1,76 @@
+package grpc.akkastreams
+
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}
+
+import akka.stream.{Attributes, FlowShape, Inlet, Outlet}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import io.grpc.stub.{ClientCallStreamObserver, ClientResponseObserver}
+
+class GrpcGraphStage[I, O](operator: GrpcOperator[I, O]) extends GraphStage[FlowShape[I, O]] {
+  val in = Inlet[I]("grpc.in")
+  val out = Outlet[O]("grpc.out")
+
+  override val shape: FlowShape[I, O] = FlowShape.of(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+
+      var requestStream = new AtomicReference[Option[ClientCallStreamObserver[I]]](None)
+      val element = new AtomicReference[Option[I]](None)
+      val requested = new AtomicBoolean(false)
```

Comment:
ok for a start, in general though you'll want to move these to integers of "how much was requested" and also make the element a buffer instead, similar to: https://github.com/akka/akka/pull/24230/files#diff-d061ab9a88e42dd22fe19bd18271b95dR75

Notice also the request strategy used in there.

I think it's good to get it working using the simple impl you're starting on here and then we can polish it and make it more perfromant (the requesting 1 by 1 will be a bit slow), so good that you started here~!

Reply:
I wonder if we can get rid of the atomics, and rather also the run() delegate via async callback and ride on the Stage's guarantee of "single threaded illusion". Would have to benchmark what's faster.

Reply:
Buffering makes sense, I'll add it.

Reply:
Probably could get rid of the Atomic using the async callbacks as you suggested. I'll give it a try.

Thanks for the feedback!